src/agg_bookend_optimize.c

```diff
@@ -0,0 +1,678 @@
+static void first_last_qp_callback(PlannerInfo *root, void *extra);
+static Node *mutate_aggref_node(Node *node, MutatorContext *context);
+static void replace_aggref_in_tlist(MinMaxAggPath *minmaxagg_path);
+
+/*
+ * mutate_aggref_node
+ *
+ * Mutator function used by recursive `expression_tree_mutator`
+ * to replace Aggref node with Param node
+ */
+Node *
+mutate_aggref_node(Node *node, MutatorContext *context)
+{
+	if (node == NULL)
+		return NULL;
+	if (IsA(node, Aggref))
+	{
+		Aggref	   *aggref = (Aggref *) node;
+
+		/* See if the Aggref should be replaced by a Param */
+		if (context->mm_path != NULL &&
+			list_length(aggref->args) == 2)
+		{
+			TargetEntry *curTarget = (TargetEntry *) linitial(aggref->args);
+			ListCell   *cell;
+
+			foreach(cell, context->mm_path->mmaggregates)
+			{
+				MinMaxAggInfo *mminfo = (MinMaxAggInfo *) lfirst(cell);
+
+				if (mminfo->aggfnoid == aggref->aggfnoid &&
+					equal(mminfo->target, curTarget->expr))
+					return (Node *) copyObject(mminfo->param);
+			}
+		}
+	}
+	return expression_tree_mutator(node, mutate_aggref_node, (void *) context);
+}
+
+/*
+ * replace_aggref_in_tlist
+ *
+ * If MinMaxAggPath is chosen, instead of running aggregate
+ * function we will execute subquery that we've generated. Since we
+ * use subquery we need to replace target list Aggref node with Param
+ * node. Param node passes output value from the subquery.
+ *
+ */
+void
+replace_aggref_in_tlist(MinMaxAggPath *minmaxagg_path)
+{
+	MutatorContext context;
+
+	context.mm_path = minmaxagg_path;
+
+	((Path *) minmaxagg_path)->pathtarget->exprs = (List *) mutate_aggref_node((Node *) ((Path *) minmaxagg_path)->pathtarget->exprs, (void *) &context);
+}
+
+static bool
+is_first_last_agg(Aggref *aggref, const char *name)
+{
+	char	   *agg_name = get_func_name(aggref->aggfnoid);
+
+	if (strncmp(agg_name, name, NAMEDATALEN) == 0)
+		return true;
+	return false;
+}
+
+static bool
+is_first_agg(Aggref *aggref)
+{
+	return is_first_last_agg(aggref, FIRST);
+}
+
+static bool
+is_last_agg(Aggref *aggref)
+{
+	return is_first_last_agg(aggref, LAST);
+}
+
+static bool
+contains_aggref_node(List *sortClause, List *targetList)
+{
+	List	   *exprs = get_sortgrouplist_exprs(sortClause, targetList);
+	ListCell   *cell;
+
+	foreach(cell, exprs)
+	{
+		Node	   *expr = lfirst(cell);
+
+		if (IsA(expr, Aggref))
+		{
+			Aggref	   *aggref = (Aggref *) expr;
+
+			if (is_first_agg(aggref) || is_last_agg(aggref))
+				return true;
+		}
+	}
+	return false;
+}
+
+/*
+ * preprocess_first_last_aggregates - preprocess FIRST/LAST aggregates
+ *
+ * Check to see whether the query contains FIRST/LAST aggregate functions that
+ * might be optimizable via index scans.  If it does, and all the aggregates
+ * are potentially optimizable, then create a MinMaxAggPath(reusing MinMax path implementation)\
+ * and add it to the (UPPERREL_GROUP_AGG, NULL) upperrel.
+ *
+ * This should be called by grouping_planner() just before it's ready to call
```

Comment:
I don't think this paragraph really applies does it?

Reply:
fixed